draft-pda-core.md

```diff
@@ -291,11 +291,437 @@ Each PA task in this document is divided into three sub-protocols as follows.
+~~~
+
+where `sk` is the secret key corresponding to the HPKE key config. For each
+report share `share`, it derives is share of the input and proof from `context`
+and `share.payload` and computes its response according to the PA protocol. It
+responds to the POST request from the leader with status 200 and the following
+message:
+
+~~~
+struct {
+  PAProto proto;
+  select (PAVerifyStartResp.proto) {
+    case prio: PrioVerifyStartResp payloads<0..2^24-1>;
+    case hits: HitsVerifyStartResp payloads<0..2^24-1>;
+  }
+} PAVerifyStartResp;
+~~~
+
+The `i`-th element of `PAVerifyStartResp.payloads` corresponds to the `i`-th element
+of `PAVerifyStartReq.payloads` for each `0 < i <= L`.
+
+### Verify Finish
+
+Next, the leader processes each message in `PAVerifyStartResp.payloads`
+according to the PA protocol and sends a request to the helper, constructed as
+follows.
+
+**Request.**
+The leader sends a POST request to `[helper]/verify_finish` with the following
+message:
+
+~~~
+struct {
+  PATask task;                     // Equal to PAVerifyStartReq.task
+  uint8 key_config_id;             // Equal to PAVerifyStartReq.key_config_id
+  PAHelperShare shares<0..2^24-1>; // Equal to PAVerifyStartReq.shares
+  select (PAVerifyFinishReq.task.proto) {
+    case prio: PrioVerifyFinishReq payloads<0..2^24-1>;
+    case hits: HitsVerifyFinishReq payloads<0..2^24-1>;
+  }
+} PAVerifyFinishReq;
+~~~
+
+As usual, the `i`-th payload corresponds to the `i`-th payload of the
+`PAVerifyStartResp` sent by the helper in response to the previous request.
+
+**Response.**
+The helper handles POST requests to `[helper]/verify_finish` as follows. It
+begins just as before by looking up the PA parameters `PAParam` for which
+`PAVerifyStartReq.task.id == PAParam.task.id`. Next, it looks up the HPKE config
+and corresponding secret key associated with `PAVerifyStartReq.key_config_id`.
+If not found, then it aborts and alerts the leader with "unrecognized key
+config".
+
+Finally, the helper responds to the POST request with status 200 and the
+following message body:
+
+~~~
+struct {
+  PAProto proto;
+  select (PAVerifyFinishResp.proto) {
+    case prio: PrioVerifyFinishResp payloads<0..2^16-1>;
+    case hits: HitsVerifyFinishResp payloads<0..2^16-1>;
+  }
+} PAVerifyFinishResp;
+~~~
+
+As usual, the `i`-th payload corresponds to the `i`-th payload of the
+`PAVerifyFinishReq`.
+
+### Decision
+
+The helper decides whether each of its input shares is valid after it responds
+to the leader's verify finish request. Likewise, the leader decides whether each
+of its shares is valid after processing the helper's response to its verify
+finish request.
+
+For each valid input, each aggregator derives its input share and stores it for
+use in [the "Collect" phase](pa-collect.md) of the PA protocol.
+
+# Collect {#pa-collect}
+
+[TODO]
+
+# Error handling {#pa-error}
+
+A protocol participant *aborts* the protocol by tearing down the TLS connection
+with each of its peers.
```

Comment:
Is this what we want to do if the helper fails to process one share of a batch?

Reply:
Sorry, I was zeroing in on the "tear down the HTTPS connection" piece. I don't think that's really required. All that seems required is for the protocol (be it upload, verify, etc) to be aborted, which might manifest in tearing down HTTPS, or it might not. For upload, aborting is likely akin to tearing down the connection, but it might not be. (Clients could, e.g., keep open persistent connections to the leader if they're uploading many things.)

Reply:
Good point! "Aborting" just means forgetting any state associated with the protocol ... it needn't result in dropping the HTTPS state altogether.

I guess we don't need to be explicit about what "abort" means right now. I just deleted this paragraph.